### PR TITLE
[BACKPORT] Attempt to better handle HTTP ReadTimeout/ConnectionError errors

### PR DESCRIPTION
**Pull request by CastagnaIT**
### Check if this PR fulfills these requirements:
* [x] My changes respect the [Kodi add-on development rules](https://kodi.wiki/view/Add-on_rules)
* [x] I have read the [**CONTRIBUTING**](../../master/Contributing.md) document.
* [x] I made sure there wasn't another [Pull Request](../../../pulls) opened for the same update/change
* [x] I have successfully tested my changes locally

#### Types of changes
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Feature change (non-breaking change which change behaviour of an existing functionality)
- [x] Improvement (non-breaking change which improve functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Description

backport of #1046

### CODE REVIEW

**Comment by jakermx** [comment]
Hi my der friend, sorry for the delay but...I am here with some findngs and solid results.

First...Excuse me if you misunderstood me about VC...I meant Virtual Circuits not Visual C....a Virtual Circutis is a Established Connecction between 2  devices, which includes Source and Destination port...so OSI Model L4 Session = VC.

Anyway...I am seeing that you tittled a Workaround as Fix...Error Handling is a Workaround since the issue persists but the developer trigger an action in order to pass it out. But as you said, the issue will be hidden and will persists.

After checking the NF playback  behaivour on Web, Android App and Roku Channel, and per your suggestion, did a deep analisys on the Addon, I found the following.

-Video Playback has a parallel session on all Devices/Software
-The Data Session(Parallel from playback) is used for Playback progress updates during playback, so keep alive sessions are used for that
-Once the playback session ends...it kills the playback session and just keep one session for MSL and Navigation
-The NF addon is using , as you said, 2 sessions for Data, MSL and Session and ISA is using and extraone
-The NF addon is mixing Sessions sometimes (MSL and NFSession)
-The requests add on is not the same as the Requests Lib, it is an Adaptation for KODI, BTW it is very old
-The NF AddOn sometimes consumes the Requests AddOn Lib, Other times the Python Lib, and some others urllib3 from scratch

Based on theese...I did some test with Requests, URLIB3, and PythonKODI...and...remrking what Documentation on Requests and urllib3 I can confirm the following.

from urllib3 site:
"These connections are used during an individual request and returned to the pool when the request is complete. By default only one connection will be saved for re-use. If you are making many requests to the same host simultaneously it might improve performance"

from Requests site:
"Note that connections are only released back to the pool for reuse once all body data has been read; be sure to either set stream to False or read the content property of the Response object."

Based on theese, I went deeper...and found that the consumming methods are not standarized, some times, consumes json,text, bytes and some others grabs a single value, and soething loks a joke...just consumes it and returns if it was ok or dont......I really look the MSLv2 I sent Jojo about a year ago , hehehe.

But now it Breaks on the  first response for chuncked responses without checking if there are data left.....on the license related methods, the challange info is fixed instead of getting it from the proxy client, anyway...after several tries and several ways to try to have real solution, I would like to suggest to have one session for data (MSL and NFSession), and to always keep sync the playback to Netflix...that way, the AddOn will have the closest behaivour to the Official Apps and will always have a Connecition open and the Serverver Timeout(10 mins) will never reached.

I have very bad and rude version of it, but a working without issues and without shortcuts or workarounds....but as you told me...I have no high level developer skills, so that why I said....it is pretty PRETTYBAD.

Well, ask me for more info or keep as it.....Cheers from Mexico  my friends

**Comment by CastagnaIT** [comment]
some points

> The requests add on is not the same as the Requests Lib, it is an Adaptation for KODI, BTW it is very old

it is not so old, Kodi v2.22 >> sources v2.25 and the changelog not specify nothing of useful things for add-on in new versions
if you think that the new version of request module could solve, you can remove the dependency from addon xml and add manually the updated request module

> But now it Breaks on the first response for chuncked responses without checking if there are data left.....

I didn't really understand this sentence

> on the license related methods, the challange info is fixed instead of getting it from the proxy client,

where?
only for the manifest request has for the first time the challenge data fixed and only for ARM devices

> Serverver Timeout(10 mins) 

again this 10-minute mention without specifying where you got this value... I have not find any reference to this
where is this timeout specified? documentations? links?

> I would like to suggest to have one session for data (MSL and NFSession)

already tried, on last test and has not worked
from the test made by user in the closed Issue thread has said it not works,
for this i have no intention of merging the two services MSL and NFSession into one for the time being,
requires a lot of changes and change like this it will not be taken into account for the Kodi 18 version

PS. and a possible "solution" must also work without "watched status sync" enabled

It also does not explain why it happens to only a few individual users, _not all users_
if it was a session problem, it should happen to everyone,
i am always suspicious that the problem may be elsewhere as this user mentition:

> i also had issues with timeout errors. i am not sure if the reason only applies to me but i found a solution.
> I "debugged" the timeout by running curl www.netflix.com and found it stuck for most of the requests. I found a blog talking about netflix blocking ipv6 tunnels which basically disables ipv6 usage. To prevent this you can "blackhole" the ipv6 address and suddently the connection always works:
> https://blog.jmwhite.co.uk/2016/06/12/netflix-starts-blocking-ipv6-tunnels/
> sudo ip -6 route add blackhole 2a05:d018:76c::/48
> 
> i hope it is of use to someone.

according to this user this solution works for him
but even if this were the case, the solution is too complex for normal users
but it would mean that the problem does not lie here

**Comment by jakermx** [comment]
> some points
> 
> > The requests add on is not the same as the Requests Lib, it is an Adaptation for KODI, BTW it is very old
> 
> it is not so old, Kodi v2.22 >> sources v2.25 and the changelog not specify nothing of useful things for add-on in new versions
> if you think that the new version of request module could solve, you can remove the dependency from addon xml and add manually the updated request module

I said old because, it hasnt been update for 2+ years,at least the KODI Addon, I was just saying

> 
> > But now it Breaks on the first response for chuncked responses without checking if there are data left.....
> 
> I didn't really understand this sentence
> 
> > on the license related methods, the challange info is fixed instead of getting it from the proxy client,
> 
> where?
> only for the manifest request has for the first time the challenge data fixed and only for ARM devices
> 

The challange is the encoded request that the end point  sends...I used to get it from the request received on the proxy server, in this case the Add On request thru the MSL Proxy Server....what really NF look for is for the Proxy Agent, Browser Version and thats it....

> > Serverver Timeout(10 mins)
> 
> again this 10-minute mention without specifying where you got this value... I have not find any reference to this
> where is this timeout specified? documentations? links?
> 

It is the time that the NF Servers waits for request on a VC, after that time  NF Servers close the VC...there is no Documentation,  you can get it on Wireshark, you will see a FIN, ACK request from the end point requesting to close the VC, in the past....it did not happen at all.


> > I would like to suggest to have one session for data (MSL and NFSession)
> 
> already tried, on last test and has not worked
> from the test made by user in the closed Issue thread has said it not works,
> for this i have no intention of merging the two services MSL and NFSession into one for the time being,
> requires a lot of changes and change like this it will not be taken into account for the Kodi 18 version
>
I know it requieres a lot of work, I just suggested
 
> PS. and a possible "solution" must also work without "watched status sync" enabled
> 
if you want to work without enabling the watched status,  the last call before playback have to have Connection Close header or....trigger a session close ...but...it is not so easy just clossing the session, because, the user could navigate in the mean time....for me...setting the Connection Close header while playback is the better and easy solution



> It also does not explain why it happens to only a few individual users, _not all users_
> if it was a session problem, it should happen to everyone,
> i am always suspicious that the problem may be elsewhere as this user mentition:
> 
> > i also had issues with timeout errors. i am not sure if the reason only applies to me but i found a solution.
> > I "debugged" the timeout by running curl [www.netflix.com](http://www.netflix.com) and found it stuck for most of the requests. I found a blog talking about netflix blocking ipv6 tunnels which basically disables ipv6 usage. To prevent this you can "blackhole" the ipv6 address and suddently the connection always works:
> > https://blog.jmwhite.co.uk/2016/06/12/netflix-starts-blocking-ipv6-tunnels/
> > sudo ip -6 route add blackhole 2a05:d018:76c::/48
> > i hope it is of use to someone.
> 
I agreed, but....what I have seen is that it only fails on never watched titles or with non exported ones

> according to this user this solution works for him
> but even if this were the case, the solution is too complex for normal users
> but it would mean that the problem does not lie here

Agreed, if it works....good, but I think it is a temp patch...and I would like to support to fix the issue from its root..

there is another workaround that developers from Requests seguessted...install a security module , but..some devices can do it and some others cant..., so I think that so far...Requests Addon is the issue, and we can redo the calls directly to urlib3, I can do it...if you share me your best practices so I complaint your programming standards.

As I have said...I like your work and effort, and I would like to support as well......


:D

**Comment by CastagnaIT** [comment]
perhaps i have not explained myself well
use a single request session for both services did not resolved the problem also with "watched status sync" enabled

this is the t15 log with MSL/NFSession services sessions unified, and the "watched status sync" enabled:
https://1drv.ms/u/s!ApvU6ekNpwo9lGnxL7bB7DGh4slS?e=e0qXaL
so maybe it solved your case but seem not the same for the others

**Comment by jakermx** [comment]
Hello,

After a lot of research, tests and personal work I found the real issue about Connection Read Timeouts and Connection Reset.

HTTP 1.1 handles the Connection : Keep-Alive by defualt, dont need to specify it, but the server could release the connection at any time based on their settings, Currently it drops a hanged connection after 10  minutes...Heres the Evidence

image
image
image


Well based on that, I did some traces and found that the when using a browser, the TCP/IP stack send a tcp keep alive packet every  45 sec after 45 sec of inactivity, but I did he same on kodi....but it doesnt  send anything, WHY!!!!

image
image
image




Because the Linux OS, URLLIB3 by default does not have it enable. so...even when the urllib3 library set the HTTP Header...the leyer 3  does not do anything, based on that I tries several ways to handle it, and finally I did this little piece of code that need to be looaded before the session creation.

    def _init_session(self):
        """Initialize the session to use for all future connections"""
        try:
            self.session.close()
            LOG.info('Session closed')
        except AttributeError:
            pass


        **import socket
        from urllib3.connection import HTTPConnection
        HTTPConnection.default_socket_options = HTTPConnection.default_socket_options + [
            (socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
            ,(socket.SOL_TCP, socket.TCP_KEEPIDLE, 45)
            ,(socket.SOL_TCP, socket.TCP_KEEPINTVL, 10)
            ,(socket.SOL_TCP, socket.TCP_KEEPCNT, 6)
             ]**

        from requests import session
        self.session = session()
.....................


image
image

And thats it!!!! now it the OS will send the correct tcp keep alive packets...until the server drops the connection or the user uses the add on before it happens.


image

As you can see...The TCP VC is still open and below...the server ask to close it instead of droping it, with out any error on the AddOn


image

This code is just for Linux Based Systems, and need to be apllied on  base.py  and msl_requests.py, talk about it, I am refeactoring the session handlers, so we can have just 1 session manager, but multi threading.. Cheers!!!

**Comment by jakermx** [comment]
I found this but I dont how to handle setsockopt and requests... maybe you know.

https://gist.github.com/shi-yan/611cc0221eeff1644797

**Comment by CastagnaIT** [comment]
i need some time to verify what you have found

on the net there is this snippet:
```

        # TCP Keep Alive Probes for different platforms
        platform = sys.platform
        # TCP Keep Alive Probes for Linux
        if platform == 'linux' and hasattr(socket, "TCP_KEEPIDLE") and hasattr(socket, "TCP_KEEPINTVL") \
                and hasattr(socket, "TCP_KEEPCNT"):
            conn.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, TCP_KEEP_IDLE)
            conn.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, TCP_KEEPALIVE_INTERVAL)
            conn.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPCNT, TCP_KEEP_CNT)

        # TCP Keep Alive Probes for Windows OS
        elif platform == 'win32' and hasattr(socket, "SIO_KEEPALIVE_VALS"):
            conn.sock.ioctl(socket.SIO_KEEPALIVE_VALS, (1, TCP_KEEP_IDLE * 1000, TCP_KEEPALIVE_INTERVAL * 1000))

        # TCP Keep Alive Probes for Mac OS
        elif platform == 'darwin':
            conn.sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
            conn.sock.setsockopt(socket.IPPROTO_TCP, TCP_KEEPALIVE, TCP_KEEPALIVE_INTERVAL)

```
could be a starting point to study how to set it for all platforms

**Comment by jakermx** [comment]
it is the same code that I found....I already sent you that link..

I think this is the solution for what urllib3 and adapters missed, becausse, urllib3 is not implementing the Keep-Alive at all...it is just sending the Connection Keep Alive by default, but...HTTP 1.1 drop it...so, if the client ask to a server for keep alive the connection,  at least do the flow control thing... but nope.

I can test the Linux and Windows thing... I dont have any Mac, and for Linux I can do it in some flavors, just share me a matrix of OS/Code and I wil fil it down for you..


:D

**Comment by jakermx** [comment]
And....another thing I found....the Proxy Servers the Add On creates....sometimes are used ...and some others nope... The idea is that the AddOn Front End  do all thru the Proxy via RPC/IPC/AddOnSignal/WebRPC-JSON.... but it is not.

And also I found a little idea about what is happenning with the authentication thing

**Comment by CastagnaIT** [comment]
i have tried do a fast check on Windows system + debugging chrome/wireshark
after several attempts i was able to see the TCP keep-alive lines
that appears to be sent with an interval of 10 seconds

i think before doing big changes,
you can test to integrate the socket fix on both web sessions NFSession/MSL
and try if works (linux only) then create a zip test and attach here

i will ask a feedback to the user on the closed Issue thread,
i prefer to have confirmation at least from another user

for windows and mac we can test later

**Comment by jakermx** [comment]
OK, I will do it...

**Comment by CastagnaIT** [comment]
from requests module is possible to make a new httpAdapter for both http/https protocols:
https://requests.readthedocs.io/en/latest/user/advanced/#example-specific-ssl-version
in this example is for SSL but this provide the access to the poolmanager class
that allow to integrate the example code in step 6 from:
https://www.finbourne.com/blog/the-mysterious-hanging-client-tcp-keep-alives

**Comment by jakermx** [comment]
Ready!

It is more code, but it is using OOP and is more classy... and overall, it is caling request


import requests
import socket
from requests.adapters import HTTPAdapter
from requests.adapters import PoolManager
from requests.packages.urllib3.connection import HTTPConnection



class SockOpsAdapter(HTTPAdapter):
  def __init__(self, options, **kwargs):
    self.options = options
    super(SockOpsAdapter, self).__init__()

  def init_poolmanager(self, connections, maxsize, block=False):
    print "init_poolmanager"
    self.poolmanager = PoolManager(num_pools=connections,
                                   maxsize=maxsize,
                                   block=block,
                                   socket_options=self.options)

 then before Session creation  use this..

-----------------------We Need to Add the OS Selector----------------
        options =  HTTPConnection.default_socket_options +  [
            (socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
            ,(socket.SOL_TCP, socket.TCP_KEEPIDLE, 45)
            ,(socket.SOL_TCP, socket.TCP_KEEPINTVL, 10)
            ,(socket.SOL_TCP, socket.TCP_KEEPCNT, 6)
             ]
-----------------------------------   Up to here   -------------------------------
        self.session  = requests.Session()


        self.session .mount('http://', SockOpsAdapter(options))
        self.session .mount('https://', SockOpsAdapter(options))


Basically it is overriding the default classes and modifiyig them as we want to....

It wold be nice if you have good fellas on RequestsPage, so , you can  ask them to implement this on their side....it happens when the client ask for Keep Alive and the Server answer with KeepAlive, Content Lenght okwithout Connection Headers.....

Also, ther is a header besides Connection that actually set the Keep alive parameters, it is called "Keep-Alive" and  has to parameters timeout and max, timeout is the idle time and max is the KA count till drop.

Anyway...let me test this on my RPI and Nano stuff


Cheers

**Comment by jakermx** [comment]
This is based on  1.12.2 for Kodi 18


[plugin.video.netflix.zip](https://github.com/CastagnaIT/plugin.video.netflix/files/5928202/plugin.video.netflix.zip)

**Comment by jakermx** [comment]
I have been tested the last Linux Solution on

NVIDIA Jetson Nano - T4L  ....I cant play videos because there is no 64 bit Lib and the T4L has no 32bit User Space, but No matter how long I dont use it....I havent received any WARNING or ERROR while Browsing

RPI 4B 8 GB  -Rasbian no RPI Oriented OS called RaspiOS, but basically it isDebian Buster, No Issues at all, it supports 1080 Playback  but need a very efficient Cooler, but about this Solution, No Error or Warning at all....I have had a same VC on it for about 2 hours, the the NF Server closed it but, it didnt affected at all

RPI 3B+ -LibreElec 9.2.6 Smooth

i3 4 RAM GB - Ubuntu Mate 18.04 LTS - Home Laptop --Smooth

i7 8 GB RAM - Ubuntu Desktop - Smooth ( Just tested a little - Itis my Job Laptop)

Xeon E5 - Debian 10 - Smooth

Centrino N3500 4GB -Ubuntu Mate 20.04 LTS - Smooth

Iam Installing on another  3 PCs Windows 7, 8.1 and 10---but as they are some kind of oldies...I dont know how loong it will take me.


But....So far so good my friend
